README.md
abcMIDI :   abc <-> MIDI conversion utilities
=============================================
midi2abc version 2.92 Aug  04 2008
abc2midi version 2.94 Oct  03 2012
abc2abc  version 1.68 Aug  08 2012
yaps     version 1.53 Aug  08 2012
abcmatch version 1.43 Aug  08 2012
midicopy version 1.10 Sep  22 2006

24th January 2002

Copyright James Allwright
J.R.Allwright (at) westminster (dot) ac (dot) uk
University of Westminster,
London, UK

July 2011

Seymour Shlien
seymour (dot) shlien (at) crc (dot) ca

This is free software. You may copy and re-distribute it under the terms of 
the GNU General Public License version 2 or later, which is available from
the Free Software Foundation (and elsewhere). 

This package is to be found on the web at

http://abc.sourceforge.net/abcMIDI/
(The latest versions for the time being is found on
http://ifdo.pugmarks.com/~seymour/runabc/top.html .)

Note, if you have difficulty compiling the package because you do not have
snprintf see the note in `doc/CHANGES` dated January 08 2005 (and also
December 17 2004).

These programs make use of the '`midifilelib`' public domain MIDI file utilities,
available from:

http://www.harmony-central.com/MIDI/midifilelib.tar.gz

If you have the source distribution and intend to re-compile the code,
read the file '`coding.txt`'.

---------------------------------------------------------------------
midi2abc - program to convert MIDI format files to abc notation. 
----------------------------------------------------------------
This program takes a MIDI format file and converts it to something as close
as possible to abc text format. The user then has to add text fields not
present in the MIDI header and possibly tidy up the abc note output.

Features :
----------
* The key is chosen so as to minimize the number of accidentals. 
Alternatively, the user can specify the key numerically (a positive number
is the number of sharps, a negative number is minus the number of flats).
* Note length can be set by specifiying the total number of bars or the 
tempo of the piece. Alternatively the note length can be read from the file.
However, by default it is deduced in a heuristic manner from the inter-note 
distances.  This means that you do not have to use the MIDI clock as a 
metronome when playing in a tune from a keyboard. 
* Barlines are automatically inserted. The user specifies the number of
measures in the anacrusis before the first barline and the time signature.
* The program can guess how many beats there should be in the anacrusis,
either by looking for the first strong note or minimizing the number of
notes split by a tie across a barline.
* Where a note extends beyond a bar break, it is split into two tied notes.
* The output has 4 bars per line.
* Enough accidental signs are put in the music to ensure that no pitch
errors occur if a barline is added or deleted.
* The program attempts to group notes sensibly in each bar.
* Triplets and broken rhythm (a&gt;b) are supported.
* Chords are identified.
* Text information from the original MIDI file is included as comments.
* The `-c` option can be used to select only 1 MIDI channel. Events on 
other channels are ignored.

What midi2abc does not do :
---------------------------
* Supply tune title, composer or any other field apart from `X:` , `K:`, `Q:`, `M:`
and `L:` - these must be added by hand afterwards, though they may have been
included in the text of the MIDI file.
* Support duplets, quadruplets, other esoteric features.
* Support mid-tune key or meter changes.
* Deduce repeats. The output is just the notes in the input file.
* Recover an abc tune as supplied to abc2midi. However, if you want to
do this, `midi2abc -xa -f file.mid` comes close.

---------------------------------------------------------------------
midi2abc 
--------
Usage :
-------
    midi2abc <options>
        -a <half L: units>
        -xa  extract anacrusis from file (find first strong note)
        -ga  guess anacrusis (minimize ties across bars)
        -gk  guess key signature by minimizing accidentals
        -gu  guess the number of midi pulses per note from note
             duration statistics in the MIDI file 
        -m <time signature>
        -b <bars wanted in output>
        -Q <tempo in quarter-notes per minute>
        -k <key signature> -6 to 6 sharps
        -c <channel>
        [-f] <input file>
        -o <output file>
        -s              do not discard very short notes
        -sr             do not notate a short rest after a note
        -sum            summary
        -nt             do not look for triplets or broken rhythm
        -u              number of midi pulses per abc time unit
        -ppu            parts per abc unit length (power of 2 only)
        -aul            denominator of abc unit length (power of 2 only)
        -obpl           one bar per line (deprecated)
        -bpl <number>   number of bars per printed line
        -bps <number>   number of bars per staff line
        -nogr           No note grouping. Space put between every note.
        -splitbars      splits bars to avoid nonhomophonic chords. 
        -splitvoices    splits voices to avoid nonhomophonic chords. 
        -midigram       No abc file is created, but a list
                        of all notes is produced. Other parameters
                        are ignored.
        -mftext         No abc file is created, but a list of all
                        the midi commands (mftext like output) is
                        produced. The output is best viewed with
                        runabc.tcl
        -ver            Prints version number and exits

Use only one of `-u`, `-gu`, `-b` and `-Q` or better none.
If none of these is present, the program uses the PPQN header
information in the MIDI file to determine the note length in
MIDI pulses. This usually results in the best output. 

The output of midi2abc is printed to the screen. To save it to a file, use
the redirection operator or the `-o` option.

e.g.
    midi2abc -f file.mid > file.abc
or
    midi2abc -f file.mid -o file.abc

By default the program uses the key signature and time signature
specified in the MIDI file. If the key signature is not specified,
the program will automatically determine the key signature by
minimizing the number of accidentals. If there is no time signature
found, then 4/4 is assumed. You may also specify a time signature
using the -m option.  Allowable time signatures are C,
4/4, 3/8, 2/4 and so on.

If the tune has an anacrusis, you should specify the number in quantum
units lengths. By default a quantum unit length is one half of
the unit length note specified by the `L:` command. (However, this
can be changed using the -ppu runtime parameter.)  For example,
if the meter is 6/8 and `L:` is set to 1/8, a half unit length is
1/16 note. An anacrusis of a quarter note would be specified as 4.

The `-bpl` and `-bps` control the formatting of the output. `-bpl` controls
the number of bars put in every line. The lines are joined together
with a backslash (`\`) until the desired number of bars per staff are
printed. Since `-bpl 1` is equivalent to `-obpl`, the latter parameter
is being deprecated.

Midi2abc has evolved quite a lot since James Allwright has
transferred support and development to me [SS]. Most of the
MIDI files posted on the web site appear to have been generated
from a music notation program and therefore convert cleanly
to abc files using all the information built into the file.
Therefore midi2abc no longer automatically estimates the number 
of MIDI pulses per note from the file statistics, but uses
the indications embedded in the file. However, the user still
has the option of calling these functions or changing this
parameter by using the run time parameters Use only one of `-u`, `-gu`, 
`-b`, and `-Q` or better none. The `-u` parameter, allows you
to specify the number of MIDI pulses per unit length, assuming
you know what it should be. To find out the value that
midi2abc is using, run the program with the `-sum` parameter.

Midi2abc quantizes the note durations to a length of half
the `L:` value. Thus if `L:` was set to 1/8, then the smallest
note that midi2abc would extract would be a 1/16 th note.
This is the basic quantum unit.  The `L:` note length is set
by midi2abc on the basis of the time signature. For time 
signatures less than 3/4 it uses an `L:` 1/16 and for others 
it uses a length of 1/8. This convention was probably choosen 
so that midi2abc does not quantize the notes to a too fine 
level producing outputs like `C2-C/4 D2-D/8` etc which would 
be difficult to read.  For some music, this may be too coarse 
and it may be preferable to allow the user take control of 
the `L:` value or allow the splitting of the `L:` value into 
smaller parts.  Two new run time parameters were introduced:
`-ppu` specifies the number of parts to split the `L:` note. 
It should be a power of 2 (eg. 1,2,4,8,16 ...) and by default 
it remains as 2 as before. The -aul specifies the `L:` value to use,
for example, to get `L: 1/32` you would put `-aul 32`.

Keyboard and guitar music has a lot of chords which frequently
poses a problem to midi2abc. If the notes in the chord do
not share a common onset or end time, midi2abc uses ties to join
the longer notes to other chords. This produces a mess looking
somewhat like:
    [AG-G-G-D-G,,,-][B/2-B/2-B/2-G/2G/2-G/2-D/2-G,,,/2-]...
which does not convert into something easy to read when
display in common music notation. Abcm2ps and abc2midi allow
a bar to be split into separate voices using the & sign.
Separating the notes which do not overlap exactly into
individual voices provides some improvement. If you encountering
this problem with your MIDI file, you may wish to try rerunning
the file with either the -splitbars or -splitvoices parameter.
This is a feature introduced in June 2005 (version 2.83).
The algorithm which separates the notes in a voice into
distinct voices (tracks) (`label_split` in `midi2abc.c`) is rather
crude and needs improvement. I welcome any suggestions
or improved code.

The `-midigram` option runs midi2abc in a completely different
mode. It does not produce any abc file but outputs a list
of all the notes found in the MIDI file. Each line of output
represents one note. For each line, the on and off time
in MIDI time units, the track number, the channel number, 
midi pitch, and midi velocity are listed. The last line
contains a single value equal to the duration of the file
in MIDI pulses.  The output is designed to eventually go into a
graphical user interface (runabc) which will display these events 
in piano roll format.

Note:
-----
midi2abc does not work correctly if the lyrics are embedded
in the same track as the notes. If you intend to run midi2abc on
the MIDI file produced by abc2midi, you need to use the `-STFW`
option if the tune contains lyrics, e.g.
    abc2midi lyric_tune.abc -STFW
Since February 01 2010, abc2midi by default places the lyrics in
the same track as the notes. See `doc/CHANGES`.

-------------------------------------------------------------------------
abc2midi  - converts abc file to MIDI file(s).
----------------------------------------------
Usage :
-------
    abc2midi <abc file> [reference number] [-c] [-v] [-o filename] [-t] [-n <value>] [-ver] [-NFNP] [-NFER] [-NGRA] [-STFW] [-NCOM] [-OCC]
        [reference number] selects a tune

CHANGES
2003 March 27:
Updated the man pages to abc2abc, abc2midi and midi2abc and created a
new man page for yaps.  Minor changes to midifile.c and midifile.h
were made to minimize the number of warning error messages. Midi2abc
was upgraded to be in line with the version I use with the script
midi2abc.tcl. The syntax in mftext.c was improved to reduce the
number of compilation warnings.


2003 April 12
abc2midi: for abc tunes with no guitar chords, an error
message "Command not recognized" is issued whenever a %%MIDI
chordprog or %%MIDI bassprog command is encountered.
The problem occurs mainly with runabc, which
performs only one pass through the tune and is unable to
know whether guitar chords are present before sending 
%%MIDI commands. Since no MIDI program command is issued
in this situation, the error message  was disabled, The
fix was made in dodeferred() in genmidi.c

2003 April 12
abc2midi assumes the ratio of 2:1 for long/short notes
in broken rhythms specified with angle brackets (eg. c>d),
Though this can be changed using the %%MIDI ratio command, it
is inconvenient for many users since it is necessary to edit
each tune in the abc file in order to effect this change.
A new command parameter -RS (ratio standard) changes default
ratio to 3:1 which is the standard music notation. The new
parameter still allows the %%MIDI ratio command to override
the default. The change was incorporated in event_init()
of store.c.
 
2003 Apri 19
The proposed abc standard allows non numeric identifications
for the voice field, eg. V: soprano instead of V:1.  If you
are creating a midi file, each unique identification string will
mapped to a sequential numbers 1,2 etc, as if you had used numbers
originally. The change allows abcMIDI to handle a number of 
new multivoiced abc files which have adopted this convention.
Only the first 20 characters of the identification
string are distinguished. Anything following the identification
string is ignored. The identification string should not end with
an equal sign =.  If nonnumeric identifications are used,
abc2midi will print the mapping between the id's and the numbers.
This may be useful for catching spelling errors in the id's.
The fix was made in function parsefield in parseabc.c. The
change does not yet allow V: commands to appear before the
body of the tune.

A number of abc files may designate a rest using x instead of
z. This means that this rest exists but should not appear 
in the printed score. I have modified the parser so it shall
now interpret x as rest; however, if you are using yaps to 
print the score, this rest will still remain visible. Abc2midi
will now treat x exactly like z. Abc2abc will probably
automatically convert the x's to z's whether you want it or not.
A minor change was made in parsemusic in parseabc.c. It is
necessary to modify toabc.c and yapstree.c to fix the
remaining problems.

2003 April 27
Guitar chords. The base note preceeded by a slash in the guitar
chord may now  be in lower case as well as upper case. For example,
"G/B" and "G/b" will both be treated the same way. This complies
with the proposed abc standard 1.7.6 (abc-draft.txt) and avoids
problems with some files in the Nottingham database. The fix was
made in event_handle_gchord in store.c.

Guitar chords. When there is a change of meter, the gchord beat
does not seem to get changed. I inserted a call to setbeat()
in writetrack() in genmidi.c whenever a TIME feature is encountered.
It was also necessary to change the definition of setbeat in
store.c from static to external.

2003 May 3
Karoake lyrics: Improved the documentation for the functions
write_syllable and getword in genmidi.c  which handle the lyrics
in the w: body line. If a bar line is placed in the lyric
line between two tied notes, for example see below,

"C6" c> G E2- | E2 (3 c d c | "E7" B> ^G E2- | E4 |
w: All of me, | * why not take | all of me.*
              ^
write_syllable loses synchrony with the music while searching
for the next bar line in the music. The problem was fixed
by resetting waitforbar to 0 between the two calls of
write_syllable, in writetrack (genmidi.c) after TNOTE is 
handled.

I found most warning messages about mismatched
lyric syllables to music notes have to do with users
failing to notate syllables correctly when tied notes
occur.  (Failure to place a * or  _ (underscore)  preceding
syllable of tied note. Confusing hyphen with underscore...)



2003 May 4

abc2midi file xref: Abc2midi has the option of converting
only a particular tune in a abcfile to a midi file by specifying
the X: reference number immediately following the abc file name.
When this feature is used, warning messages such as

Warning in line 28: T: outside tune body - possible missing X:

The problem occurs in the function parsefield in parseabc.c.
In order to catch the reference number of the selected tune,
every line of the input file is parsed. A flag "parsing" equal
to 0 or 1 is used to turn on the parsing inside the body of
the of the music. This flag is turned in function event_refno()
in store.c by calling the function parseon() in parseabc.c.
Unfortunately all field lines (eg. T: , M:, L:, V: ...) were 
unaffected by this flag and continued to be parsed. When the
T: field command is encountered, it falls through to the
default: in parsefield and it is processed by the function,
event_field in store.c. Event_field was designed to treat
the T: and R: conditions but before handling these conditions
it checks whether the control flag dotune is set. This flag
is also set by event_refno. If the flag is not set, then 
event_field assumes that a T: field command was encountered
without a preceding X: reference field which is a valid

queues.c
/*
 * abc2midi - program to convert abc files to MIDI files.
 * Copyright (C) 1999 James Allwright
 * e-mail: J.R.Allwright@westminster.ac.uk
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

/* queues.c
 * This file is part of the code for abc2midi.
 * Notes due to finish in the future are held in a queue (linked list)
 * in time order. Qhead points to the head of the list and addtoQ() 
 * adds a note to the list. The unused elements of array Q are held
 * in another linked list pointed to by freehead. The tail is pointed
 * to by freetail. removefromQ() removes an element (always from the
 * head of the list) and adds it to the free list. Qinit() initializes
 * the queue and clearQ() outputs all the remaining notes at the end
 * of a track.
 * Qcheck() and PrintQ() are diagnostic routines.
 */

#include <stdio.h>
#include <stdlib.h>
#include "queues.h"
#include "abc.h"
#include "genmidi.h"
#include "midifile.h"

void write_event_with_delay(int delta,int event_type,int channel,char* data,int n);

/* queue for notes waiting to end */
/* allows us to do general polyphony */
#define QSIZE 50
struct Qitem {
  int delay;
  int pitch;
  int chan;
  int effect;  /* [SS] 2012-12-11 */
  int next;
};
struct Qitem Q[QSIZE+1];
int Qhead, freehead, freetail;
extern int totalnotedelay; /* from genmidi.c [SS] */
extern int notedelay;      /* from genmidi.c [SS] */
extern int bendvelocity;   /* from genmidi.c [SS] */
extern int bendacceleration; /* from genmidi.c [SS] */
extern int bendstate; /* from genmidi.c [SS] */
/* [SS] 2014-09-10 */
extern int benddata[256]; /* from genmidi.c [SS] 2015-09-10 2015-10-03 */
extern int bendnvals;
extern int controldata[3][256]; /* extended to 256 2015-10-03 */
extern int controlnvals[2];
extern int controldefaults[128]; /* [SS] 2015-08-10 */
extern int nlayers; /* [SS] 2015-08-20 */

void set_control_defaults() {
    int i;
    for (i=0;i<128;i++) controldefaults[i] = 0;
    controldefaults[7] = 96; /* volume coarse */
    controldefaults[8] = 64;  /* balance coarse */
    controldefaults[11] = 127; /* expression */
    controldefaults[10] = 64; /* pan position coarse */
}

/* routines to handle note queue */

/* genmidi.c communicates with queues.c mainly through the    */
/* functions addtoQ and timestep. The complexity comes in the */
/* handling of chords. When another note in a chord is passed,*/
/* addtoQ detemines whether other notes in the Q structure    */
/* overlap in time with this chord and modifies the delay item*/
/* of the note which finish later so that it is relative to the*/
/* end of the earlier note. Normally all notes in the chord end*/
/* at the same as specifiedy abc standard, so the delay of the*/
/* other notes cached in the Q structure should be set to zero.*/

void addtoQ(num, denom, pitch, chan, effect, d)
int num, denom, pitch, chan, d;
int effect; /* [SS] 2012-12-11 */
{
  int i, done;
  int wait;
  int *ptr;

  wait = ((div_factor*num)/denom) + d;
  /* find free space */
  if (freehead == -1) {
    /* printQ(); */
    event_error("Too many notes in chord - probably missing ']' or '+'");
    return;
  } else {
    i = freehead;
    freehead = Q[freehead].next;
  };
  Q[i].pitch = pitch;
  Q[i].chan = chan;
  Q[i].effect = effect;  /* [SS] 2012-12-11 */
  /* find place in queue */
  ptr = &Qhead;
  done = 0;
  while (!done) {
    if (*ptr == -1) {
      *ptr = i;
      Q[i].next = -1;
      Q[i].delay = wait;
      done = 1;
    } else {
      if (Q[*ptr].delay > wait) {
        Q[*ptr].delay = Q[*ptr].delay - wait -notedelay;
        if (Q[*ptr].delay < 0) Q[*ptr].delay = 0;
        Q[i].next = *ptr;
        Q[i].delay = wait;
        *ptr = i;
        done = 1;
      } else {
        wait = wait - Q[*ptr].delay;
        ptr = &Q[*ptr].next;
      };
    };
  };
}

void removefromQ(i)
int i;
{
  if (i == -1) {
    printQ();
    event_fatal_error("Internal error - nothing to remove from queue");
  } else {
    if (Q[Qhead].delay != 0) {
    printQ();
      event_fatal_error("Internal error - queue head has non-zero time");
    };
    Qhead = Q[i].next;
    Q[i].next = freehead;
    freehead = i;
  };
}

void clearQ()
{
  int time;
  int i;

  /* remove gchord requests */
  time = 0;
  while ((Qhead != -1) && (Q[Qhead].pitch == -1)) {
    time = time + Q[Qhead].delay;
    i = Qhead;
    Qhead = Q[i].next;
    Q[i].next = freehead;
    freehead = i;
  };
  if (Qhead != -1) {
    timestep(time, 1);
  };
  /* do any remaining note offs, but don't do chord request */
  while (Qhead != -1) {
    event_error("Sustained notes beyond end of track");
    timestep(Q[Qhead].delay+1, 1);
  };
timestep(25,0); /* to avoid transient artefacts at end of track */
}

void printQ()
{
  int t;

  printf("Qhead = %d freehead = %d freetail = %d\n", 
         Qhead, freehead, freetail);
  t = Qhead;
  printf("Q:");
  while (t != -1) {
    printf("p(%d)-%d->", Q[t].pitch, Q[t].delay);
    t = Q[t].next;
  };
  printf("\n");
}

void advanceQ(t)
int t;
{
  if (Qhead == -1) {
    event_error("Internal error - empty queue");
  } else {
    Q[Qhead].delay = Q[Qhead].delay - t;
  };
}

void Qinit()
{
  int i;

  /* initialize queue of notes waiting to finish */
  Qhead = -1;
  freehead = 0;
  for (i=0; i<QSIZE-1; i++) {
    Q[i].next = i + 1;
  };
  Q[QSIZE-1].next = -1;
  freetail = QSIZE-1;
}

void Qcheck()
{
  int qfree, qused;
  int nextitem;
  int used[QSIZE];
  int i;
  int failed;

  failed = 0;
  for (i=0; i<QSIZE; i++) {
    used[i] = 0;
  };
  qused = 0;
  nextitem = Qhead;
  while (nextitem != -1) {
    qused = qused + 1;
    used[nextitem] = 1;
    nextitem = Q[nextitem].next;
    if ((nextitem < -1) || (nextitem >= QSIZE)) {
      failed = 1;
      printf("Queue corrupted Q[].next = %d\n", nextitem);
    };
  };
  qfree = 0;
  nextitem = freehead;
  while (nextitem != -1) {
    qfree = qfree + 1;
    used[nextitem] = 1;
    nextitem = Q[nextitem].next;
    if ((nextitem < -1) || (nextitem >= QSIZE)) {
      failed = 1;
      printf("Free Queue corrupted Q[].next = %d\n", nextitem);
    };
  };
  if (qfree + qused < QSIZE) {
    failed = 1;
    printf("qfree = %d qused = %d\n", qused, qfree);
  };
  for (i=0; i<QSIZE; i++) {
    if (used[i] == 0) {
      printf("Not used element %d\n", i);
      failed = 1;
    };
  };
  if (Q[freetail].next != -1) {
    printf("freetail = %d, Q[freetail].next = %d\n", freetail, 
           Q[freetail].next);
  };
  if (failed == 1) {
    printQ();
    event_fatal_error("Qcheck failed");
  };
}


/* [SS] 2012-12-11 */
void note_effect() {
/* note_effect is used only if !bend! is called without
   calling %%MIDI bendstring or %%MIDI bendvelocity [SS] 2015-08-11 */
/* Bends a note along the shape of a parabola. The note is
   split into 8 segments. Given the bendacceleration and
   initial bend velocity, the new pitch bend is computed
   for each time segment.
*/
  int delta8;
  int pitchbend;
  char data[2];
  int i;
  int velocity;
  delta8 = delta_time/8;
  pitchbend = bendstate;  /* [SS] 2014-09-09 */
  velocity = bendvelocity;
  for (i=0;i<8;i++) {
     pitchbend = pitchbend + velocity;
     velocity = velocity + bendacceleration;
     if (pitchbend > 16383) pitchbend = 16383;
     if (pitchbend < 0) pitchbend = 0;
 
     data[0] = (char) (pitchbend&0x7f);
     data[1] = (char) ((pitchbend>>7)&0x7f);
     mf_write_midi_event(delta8,pitch_wheel,Q[Qhead].chan,data,2);
     delta_time -= delta8;
     }
  midi_noteoff(delta_time, Q[Qhead].pitch, Q[Qhead].chan);
  pitchbend = bendstate; /* [SS] 2014-09-22 */
  data[0] = (char) (pitchbend&0x7f);
  data[1] = (char) ((pitchbend>>7)&0x7f);
  mf_write_midi_event(delta_time,pitch_wheel,Q[Qhead].chan,data,2);
  }

/* [SS] 2014-09-11 */
void note_effect2() {
  /* implements %%MIDI bendstring command. The note is split
     into n segments where n is the number of pitch shift values
     supplied in bendstring. For every segment, the pitchbend
     is shifted by the next pitch shift value.
  */
  int delta;
  int pitchbend;
  char data[2];
  int i;
  delta = delta_time/bendnvals;
  pitchbend = bendstate;  /* [SS] 2014-09-09 */
  for (i=0;i<bendnvals;i++) {
     pitchbend = pitchbend + benddata[i];
     if (pitchbend > 16383) pitchbend = 16383;
     if (pitchbend < 0) pitchbend = 0;
 
     data[0] = (char) (pitchbend&0x7f);
     data[1] = (char) ((pitchbend>>7)&0x7f);
     if (i == 0) /* [SS] 2014-09-24 */
       mf_write_midi_event(0,pitch_wheel,Q[Qhead].chan,data,2);
     else {
       mf_write_midi_event(delta,pitch_wheel,Q[Qhead].chan,data,2);
       delta_time -= delta;
       }
     }
  midi_noteoff(delta_time, Q[Qhead].pitch, Q[Qhead].chan);
  delta_time = 0; /* [SS] 2014-09-24 */
  pitchbend = bendstate; /* [SS] 2014-09-09 */
  data[0] = (char) (pitchbend&0x7f);
  data[1] = (char) ((pitchbend>>7)&0x7f);
  mf_write_midi_event(delta_time,pitch_wheel,Q[Qhead].chan,data,2);
  }

/* [SS] 2014-09-22 */
void note_effect3() {
  /* handles the %%MIDI bendstring command when only one pitch
     shift is given.
  */
  int delta;
  int pitchbend;
  char data[2];
  delta =0;
  pitchbend = bendstate; 
  pitchbend = pitchbend + benddata[0];
  if (pitchbend > 16383) pitchbend = 16383;
  if (pitchbend < 0) pitchbend = 0;
  data[0] = (char) (pitchbend&0x7f);
  data[1] = (char) ((pitchbend>>7)&0x7f);
  mf_write_midi_event(delta,pitch_wheel,Q[Qhead].chan,data,2);
  midi_noteoff(delta_time, Q[Qhead].pitch, Q[Qhead].chan);
  pitchbend = bendstate;
  data[0] = (char) (pitchbend&0x7f);
  data[1] = (char) ((pitchbend>>7)&0x7f);
  mf_write_midi_event(delta,pitch_wheel,Q[Qhead].chan,data,2); /* [SS] 2014-09-23 */
  } 

/* [SS] 2015-07-26 */
void note_effect4(chan)
    int chan;
    {
    /* Like %%MIDI bendstring, this procedure handles %%MIDI controlstring
       command. The first value of controlstring indicates the controller
       to modify. The following values are the values to send to the
       controller for each note segment.

       This procedure is no longer used. Presently note_effect5()
       has taken over this operation.
    */
    int delta;
    char data[2];
    int controltype,controlval;
    int i;
    if (controlnvals[0] <= 1) {
        event_error("missing %%MIDI controlstring ...");
        return;
        }
    delta = delta_time/(controlnvals[0]-1);
    controltype = controldata[0][0];
    if (controltype > 127 || controltype < 0) {
        event_error("control code muste be in range 0 to 127");
        return;
       }

    for (i = 1;i<controlnvals[0];i++)
        {
        controlval = controldata[0][i]; 
        if (controlval >127 || controlval < 0) {
            event_error("control code muste be in range 0 to 127");
            controlval = 0;            
            }
        data[0] = controltype;
        data[1] = controlval;
        write_event_with_delay(delta,control_change,chan,data,2);
        delta_time -= delta;
        }
    midi_noteoff(delta_time, Q[Qhead].pitch, Q[Qhead].chan);
    }


/* [SS] 2015-07-30 */
struct eventstruc {int time;
                   char cmd;
                   char data1;
                   char data2;};

int int_compare_events(const void *a, const void *b) {
   struct eventstruc *ia  = (struct eventstruc *)a;
   struct eventstruc *ib  = (struct eventstruc *)b;
   if (ib->time > ia->time) 
       return -1;
   else if (ia ->time > ib->time)
       return  1;
   else return 0; 
   }

void print_eventlist(struct eventstruc *list, int nsize) {
    int i;
    for (i = 0; i<nsize; i++) 
        printf("%d %d %d %d\n",list[i].time,list[i].cmd,list[i].data1,list[i].data2);
    }

/* [SS] 2015-08-03 */
void output_eventlist (struct eventstruc *list, int nsize, int chan) {
    int i;
    int delta;
    char cmd,data[2];
    int miditime;
    miditime = 0;
    for (i = 0; i<nsize; i++) {
        delta = list[i].time - miditime;
        miditime = list[i].time;
        data[0] = list[i].data1; 
        data[1] = list[i].data2;
        cmd     = list[i].cmd;
        if (cmd == -80) write_event_with_delay(delta,control_change,chan,data,2);
        if (cmd == -32) mf_write_midi_event(delta,pitch_wheel,Q[Qhead].chan,data,2);
        }
    } 

/* [SS] 2015-08-01 */
void note_effect5(chan)
    int chan;
    {
    /* This procedure merges the controlstring with the
       bendstring and uses it to shape the note. The control
       commands are prepared and stored in the eventstruc
       eventlist[] array, sorted chronologically and sent
       to the MIDI file.
    */
    struct eventstruc eventlist[1024]; /* extended to 1000 2015-10-03 */
    int delta,notetime,pitchbend;  
    int initial_bend; /* [SS] 2015-08-25 */
    int i,j;
    int layer;
    int controltype,controlval;
    char data[2];

    j = 0; 
    pitchbend = bendstate;
    initial_bend = bendstate;
    if (bendnvals > 0) {
        delta = delta_time/bendnvals;
        notetime = 0;
        for (i = 0; i <bendnvals; i++) {
            pitchbend = benddata[i] + pitchbend;
            if (pitchbend > 16383) {
               event_error("pitchbend exceeds 16383");
               pitchbend = 16383;
               }
            if (pitchbend < 0) {
               event_error("pitchbend is less than 0");
               pitchbend = 0;
               }
            eventlist[j].time = notetime;
            eventlist[j].cmd = pitch_wheel;
            eventlist[j].data1 = pitchbend & 0x7f;
            eventlist[j].data2 = (pitchbend >> 7) & 0x7f;
            notetime += delta;
            j++;
            } 
      }
 
    for (layer=0;layer <= nlayers;layer++) {
        if (controlnvals[layer] > 1) {
            delta = delta_time/(controlnvals[layer] -1);
            notetime = 0;
            controltype = controldata[layer][0];
            if (controltype > 127 || controltype < 0) {
                event_error("controller must be in range 0 to 127");
                return;
                }
    
            for (i = 1; i <controlnvals[layer]; i++) {
                controlval = controldata[layer][i];
                if (controlval < 0) {
                    event_error("control data must be zero or greater");
                    controlval = 0;
                    }
                if (controlval > 127) {
                    event_error("control data must be less or equal to 127");
                    controlval = 127;
                    }
                eventlist[j].time = notetime;
                eventlist[j].cmd  = control_change;
                eventlist[j].data1 = controltype;
                eventlist[j].data2 = controlval;
                notetime += delta;
                if (j < 1023) j++; /* [SS] increased to 999 2015-10-04 */
                else event_error("eventlist too complex");
               
                }
            }         
        }

    /* [SS] 2015-08-23 */
    if (j > 1) {
        qsort(eventlist,j,sizeof(struct eventstruc),int_compare_events);
   /* print_eventlist(eventlist,j); */
        }
    output_eventlist(eventlist,j,chan);

    /* [SS] 2015-08-28 */
    midi_noteoff(delta, Q[Qhead].pitch, Q[Qhead].chan);

    for (layer=0;layer <= nlayers;layer++) {
        controltype = controldata[layer][0];
        data[0] = (char) controltype;
        data[1] = (char) controldefaults[controltype];
        write_event_with_delay(0,control_change,chan,data,2);
        }
    if (bendnvals > 0) {
      /* restore pitchbend to its original state */
      pitchbend = initial_bend; /* [SS] 2014-09-25 */
      data[0] = (char) (pitchbend&0x7f);
      data[1] = (char) ((pitchbend>>7)&0x7f);
      write_event_with_delay(0,pitch_wheel,chan,data,2);
      }
   }




/* timestep is called by delay() in genmidi.c typically at the */
/* end of a note, chord or rest. It is also called by clearQ in*/
/* this file. Timestep, is not only responsible for sending the*/
/* midi_noteoff command for any expired notes in the Q structure*/
/* but also maintains the delta_time global variable which is  */
/* shared with genmidi.c. Timestep also calls advanceQ() in   */
/* this file which updates all the delay variables for the items */
/* in the Q structure to reflect the current MIDI time. Timestep */
/* also calls removefromQ in this file which cleans out expired */
/* notes from the Q structure. To make things even more complicated*/
/* timestep runs the dogchords and the dodrums for bass/chordal */
/* and drum accompaniments by calling the function progress_sequence.*/
/* Dogchords and dodrums may also call addtoQ changing the contents*/
/* of the Q structure array. The tracklen variable in MIDI time */
/* units is also maintained here.                               */ 

/* new: delta_time_track0 is declared in queues.h like delta_time */

void timestep(t, atend)
int t;
int atend;
{
  int time;
  int headtime;

  time = t;
  /* process any notes waiting to finish */
  while ((Qhead != -1) && (Q[Qhead].delay < time)) {
    headtime = Q[Qhead].delay;
    delta_time = delta_time + (long) headtime;
    delta_time_track0 = delta_time_track0 + (long) headtime; /* [SS] 2010-06-27*/
    time = time - headtime;
    advanceQ(headtime);
    if (Q[Qhead].pitch == -1) {
      if (!atend) {
        progress_sequence(Q[Qhead].chan);
      };
    } else {
       if (Q[Qhead].effect == 0) {
          midi_noteoff(delta_time, Q[Qhead].pitch, Q[Qhead].chan);
          tracklen = tracklen + delta_time;
          delta_time = 0L;}
       else {
          tracklen = tracklen + delta_time; /* [SS] 2015-06-08 */
          switch (Q[Qhead].effect) { /* [SS] 2014-09-22 */
             case 1:
                note_effect();
                break;
   
             case 2:
                note_effect2();
                break;

             case 3:
                note_effect3();
                break;

             case 10:
                /*note_effect4(Q[Qhead].chan);*/
                note_effect5(Q[Qhead].chan);
                break;
           } 
          delta_time = 0L;}
       };

    removefromQ(Qhead);
  };
  if (Qhead != -1) {
    advanceQ(time);
  };
  delta_time = delta_time + (long)time - totalnotedelay;
  delta_time_track0 = delta_time_track0 + (long)time - totalnotedelay; /* [SS] 2010-06-27*/
}

